Kitodo-FileManagement/src/main/java/org/kitodo/filemanagement/locking/LockManagement.java

```diff
@@ -0,0 +1,511 @@
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.kitodo.api.filemanagement.LockingMode;
+import org.kitodo.api.filemanagement.LockingResult;
+
+/**
+ * This is the main class of lock management. In regular operation, there must
+ * always be just one lock management for comprehensible reasons, for which the
+ * class implements the singleton pattern. The lock management has been divided
+ * into three classes for the sake of clarity, the actual lock management is in
+ * this class and manages the locks as such. The stream management manages the
+ * input and output streams flowing through the locks, and the immutable read
+ * file management manages the temporary files that must be generated for the
+ * immutable read lock and later deleted.
+ */
+public class LockManagement {
+    /**
+     * The singleton instance of the lock management. This class should be used
+     * as singleton, meaning that there is only one lock management all over the
+     * class loader. Everything else would not make sense or would even be
+     * dangerous. Nevertheless, the methods were not simply implemented
+     * {@code static}ally, because in the tests for this class, lock management
+     * must be re-initialized for each test. But for the usage, only the
+     * singleton should be used and no more instances of the class should exist.
+     */
+    private static volatile LockManagement instance;
+
+    private static final Logger logger = LogManager.getLogger(LockManagement.class);
+
+    /**
+     * This message is used in exceptions when a wrong object has been
+     * submitted.
+     */
+    private static final String UNSUITABLE_LOCKING_RESULT = "The locking result is unsuitable for generating the stream: ";
+
+    /**
+     * Returns the singleton instance of the lock management. This class should
+     * be used as singleton, meaning that there is only one lock management all
+     * over the class loader. For the usage, only the singleton should be used
+     * and no more instances of the class should exist. Everything else would
+     * not make sense or would even be dangerous.
+     *
+     * @return the singleton instance of this class
+     *
+     * @see "https://en.wikipedia.org/wiki/Double-checked_locking"
+     */
+    public static LockManagement singleton() {
+        if (instance == null) {
+            synchronized (LockManagement.class) {
+                if (instance == null) {
+                    instance = new LockManagement();
+                }
+            }
+        }
+        return instance;
+    }
+
+    private static GrantedAccess tryCast(LockingResult lockingResult) throws AccessDeniedException {
+        if (lockingResult == null || !(lockingResult instanceof GrantedAccess)) {
```

Comment:
Unnecessary null check before instanceof - instanceof returns false in case of null.

Reply:
I did not know that yet! Great! 👍